File: ios/Sources/EncoreCapacitorPlugin/EncoreCapacitorPlugin.swift
// EncoreCapacitorPlugin.swift
// iOS bridge — delegates all calls to the native encore-swift-sdk.

import Foundation
import Capacitor
import Encore

@objc(EncoreCapacitorPlugin)
public class EncoreCapacitorPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "EncoreCapacitorPlugin"
    public let jsName = "EncoreCapacitorPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "identify", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserAttributes", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reset", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setClaimEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "registerCallbacks", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "completePurchaseRequest", returnType: CAPPluginReturnPromise),
    ]

    private var currentContinuation: CheckedContinuation<Void, Error>?
    private let lock = NSLock()

    // MARK: - Configuration

    @objc func configure(_ call: CAPPluginCall) {
        guard let apiKey = call.getString("apiKey") else {
            call.reject("Missing required parameter: apiKey")
            return
        }

        let logLevel: Encore.LogLevel
        switch call.getString("logLevel") {
        case "error": logLevel = .error
        case "warn": logLevel = .warn
        case "info": logLevel = .info
        case "debug": logLevel = .debug
        default: logLevel = .none
        }

        DispatchQueue.main.async {
            Encore.shared.configure(apiKey: apiKey, logLevel: logLevel)
            call.resolve(["success": true])
        }
    }

    // MARK: - User Identity

    @objc func identify(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject("Missing required parameter: userId")
            return
        }

        let attrs = Self.parseAttributes(call.getObject("attributes"))
        Encore.shared.identify(userId: userId, attributes: attrs)
        call.resolve(["success": true])
    }

    @objc func setUserAttributes(_ call: CAPPluginCall) {
        guard let attrsObj = call.getObject("attributes") else {
            call.reject("Missing required parameter: attributes")
            return
        }

        if let attrs = Self.parseAttributes(attrsObj) {
            Encore.shared.setUserAttributes(attrs)
        }
        call.resolve(["success": true])
    }

    @objc func reset(_ call: CAPPluginCall) {
        cancelStalePurchase()
        Encore.shared.reset()
        call.resolve(["success": true])
    }

    // MARK: - Claim Control

    @objc func setClaimEnabled(_ call: CAPPluginCall) {
        guard let enabled = call.getBool("enabled") else {
            call.reject("Missing required parameter: enabled")
            return
        }

        Encore.shared.placements.isClaimEnabled = enabled
        call.resolve(["success": true])
    }

    // MARK: - Offers

    @objc func show(_ call: CAPPluginCall) {
        let placementId = call.getString("placementId")

        cancelStalePurchase()
        Task {
            do {
                let result = try await Encore.placement(placementId).show()
                switch result {
                case .granted(let entitlement):
                    call.resolve(["status": "granted", "entitlement": "\(entitlement)"])
                case .notGranted(let reason):
                    call.resolve(["status": "not_granted", "reason": reason.rawValue])
                }
            } catch {
                call.reject("SHOW_FAILED", error.localizedDescription, error)
            }
        }
    }

    // MARK: - Callbacks

    @objc func registerCallbacks(_ call: CAPPluginCall) {
        Encore.shared
            .onPurchaseRequest { [weak self] request in
                guard let self = self else { return }

                self.cancelStalePurchase()

                try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
                    self.lock.lock()
                    self.currentContinuation = continuation
                    self.lock.unlock()

                    self.notifyListeners("onPurchaseRequest", data: [
                        "productId": request.productId,
                        "placementId": request.placementId as Any,
                        "promoOfferId": request.promoOfferId as Any,
                    ])
                }
            }
            .onPurchaseComplete { [weak self] transaction, productId in
                guard let self = self else { return }
                self.notifyListeners("onPurchaseComplete", data: [
                    "productId": productId,
                    "transactionId": "\(transaction.id)",
                ])
            }
            .onPassthrough { [weak self] placementId in
                guard let self = self else { return }
                self.notifyListeners("onPassthrough", data: [
                    "placementId": placementId as Any,
                ])
            }

        call.resolve(["success": true])
    }

    @objc func completePurchaseRequest(_ call: CAPPluginCall) {
        let success = call.getBool("success") ?? false

        lock.lock()
        let continuation = currentContinuation
        currentContinuation = nil
        lock.unlock()

        if let continuation {
            if success {
                continuation.resume()
            } else {
                continuation.resume(throwing: NSError(
                    domain: "EncoreCapacitorPlugin",
                    code: -2,
                    userInfo: [NSLocalizedDescriptionKey: "Purchase failed by JS handler"]
                ))
            }
            call.resolve(["success": true])
        } else {
            call.resolve(["success": false, "error": "No pending purchase request"])
        }
    }

    // MARK: - Internal

    private func cancelStalePurchase() {
        lock.lock()
        let stale = currentContinuation
        currentContinuation = nil
        lock.unlock()

        stale?.resume(throwing: NSError(
            domain: "EncoreCapacitorPlugin",
            code: -3,
            userInfo: [NSLocalizedDescriptionKey: "Superseded by new SDK action"]
        ))
    }

    // MARK: - Helpers

    private static func parseAttributes(_ dict: [String: Any]?) -> UserAttributes? {
        guard let dict = dict else { return nil }
        return UserAttributes(
            email: dict["email"] as? String,
            firstName: dict["firstName"] as? String,
            lastName: dict["lastName"] as? String,
            phoneNumber: dict["phoneNumber"] as? String,
            postalCode: dict["postalCode"] as? String,
            city: dict["city"] as? String,
            state: dict["state"] as? String,
            countryCode: dict["countryCode"] as? String,
            latitude: dict["latitude"] as? String,
            longitude: dict["longitude"] as? String,
            dateOfBirth: dict["dateOfBirth"] as? String,
            gender: dict["gender"] as? String,
            language: dict["language"] as? String,
            subscriptionTier: dict["subscriptionTier"] as? String,
            monthsSubscribed: dict["monthsSubscribed"] as? String,
            billingCycle: dict["billingCycle"] as? String,
            lastPaymentAmount: dict["lastPaymentAmount"] as? String,
            lastActiveDate: dict["lastActiveDate"] as? String,
            totalSessions: dict["totalSessions"] as? String,
            custom: dict["custom"] as? [String: String] ?? [:]
        )
    }
}

File: android/src/main/java/com/encorekit/capacitor/EncoreCapacitorPlugin.kt
// EncoreCapacitorPlugin.kt
// Android bridge — delegates all calls to the native encore-android-sdk.

package com.encorekit.capacitor

import com.encorekit.encore.Encore
import com.encorekit.encore.core.canonical.user.UserAttributes
import com.encorekit.encore.core.infrastructure.logging.LogLevel
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "EncoreCapacitorPlugin")
class EncoreCapacitorPlugin : Plugin() {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    @Volatile private var currentDeferred: CompletableDeferred<Boolean>? = null

    // -- Configuration --

    @PluginMethod
    fun configure(call: PluginCall) {
        val apiKey = call.getString("apiKey")
        if (apiKey == null) {
            call.reject("Missing required parameter: apiKey")
            return
        }

        val logLevel = when (call.getString("logLevel")) {
            "error" -> LogLevel.ERROR
            "warn" -> LogLevel.WARN
            "info" -> LogLevel.INFO
            "debug" -> LogLevel.DEBUG
            else -> LogLevel.NONE
        }

        Encore.shared.configure(context, apiKey = apiKey, logLevel = logLevel)
        call.resolve(JSObject().put("success", true))
    }

    // -- User Identity --

    @PluginMethod
    fun identify(call: PluginCall) {
        val userId = call.getString("userId")
        if (userId == null) {
            call.reject("Missing required parameter: userId")
            return
        }

        Encore.shared.identify(userId = userId, attributes = parseAttributes(call))
        call.resolve(JSObject().put("success", true))
    }

    @PluginMethod
    fun setUserAttributes(call: PluginCall) {
        parseAttributes(call)?.let { Encore.shared.setUserAttributes(it) }
        call.resolve(JSObject().put("success", true))
    }

    @PluginMethod
    fun reset(call: PluginCall) {
        Encore.shared.reset()
        call.resolve(JSObject().put("success", true))
    }

    // -- Claim Control --

    @PluginMethod
    fun setClaimEnabled(call: PluginCall) {
        val enabled = call.getBoolean("enabled")
        if (enabled == null) {
            call.reject("Missing required parameter: enabled")
            return
        }

        Encore.shared.placements.isClaimEnabled = enabled
        call.resolve(JSObject().put("success", true))
    }

    // -- Offers --

    @PluginMethod
    fun show(call: PluginCall) {
        val placementId = call.getString("placementId") ?: ""
        val activity = activity
        if (activity == null) {
            call.reject("NO_ACTIVITY", "No current activity available")
            return
        }

        scope.launch {
            try {
                val result = Encore.shared.placement(placementId).show(activity)
                val obj = JSObject()
                when (result) {
                    is com.encorekit.encore.features.offers.PresentationResult.Completed -> {
                        obj.put("status", "completed")
                        obj.put("offerId", result.offerId)
                        obj.put("campaignId", result.campaignId)
                    }
                    is com.encorekit.encore.features.offers.PresentationResult.Dismissed -> {
                        obj.put("status", "dismissed")
                        obj.put("reason", result.reason.value)
                    }
                    is com.encorekit.encore.features.offers.PresentationResult.NoOffers -> {
                        obj.put("status", "no_offers")
                    }
                }
                call.resolve(obj)
            } catch (e: Exception) {
                call.reject("SHOW_FAILED", e.message, e)
            }
        }
    }

    // -- Callbacks --

    @PluginMethod
    fun registerCallbacks(call: PluginCall) {
        Encore.shared
            .onPurchaseRequest { request ->
                // Fail stale deferred so the previous purchase flow resolves
                currentDeferred?.complete(false)

                val deferred = CompletableDeferred<Boolean>()
                currentDeferred = deferred

                try {
                    notifyListeners("onPurchaseRequest", JSObject().apply {
                        put("productId", request.productId)
                        put("placementId", request.placementId)
                        put("promoOfferId", request.promoOfferId)
                    })
                } catch (_: Exception) {
                    // notifyListeners can throw during context teardown — safe to ignore
                }

                val success = deferred.await()
                currentDeferred = null

                if (!success) {
                    throw Exception("Purchase failed by JS handler")
                }
            }
            .onPurchaseComplete { result, productId ->
                notifyListeners("onPurchaseComplete", JSObject().apply {
                    put("productId", productId)
                    put("purchaseToken", result.purchaseToken)
                    put("orderId", result.orderId)
                })
            }
            .onPassthrough { placementId ->
                notifyListeners("onPassthrough", JSObject().apply {
                    put("placementId", placementId)
                })
            }

        call.resolve(JSObject().put("success", true))
    }

    @PluginMethod
    fun completePurchaseRequest(call: PluginCall) {
        val success = call.getBoolean("success") ?: false
        val deferred = currentDeferred
        if (deferred != null) {
            deferred.complete(success)
            call.resolve(JSObject().put("success", true))
        } else {
            call.resolve(JSObject().apply {
                put("success", false)
                put("error", "No pending purchase request")
            })
        }
    }

    // -- Helpers --

    private fun parseAttributes(call: PluginCall): UserAttributes? {
        val obj = call.getObject("attributes") ?: return null
        return UserAttributes(
            email = obj.optString("email", null),
            firstName = obj.optString("firstName", null),
            lastName = obj.optString("lastName", null),
            phoneNumber = obj.optString("phoneNumber", null),
            postalCode = obj.optString("postalCode", null),
            city = obj.optString("city", null),
            state = obj.optString("state", null),
            countryCode = obj.optString("countryCode", null),
            latitude = obj.optString("latitude", null),
            longitude = obj.optString("longitude", null),
            dateOfBirth = obj.optString("dateOfBirth", null),
            gender = obj.optString("gender", null),
            language = obj.optString("language", null),
            subscriptionTier = obj.optString("subscriptionTier", null),
            monthsSubscribed = obj.optString("monthsSubscribed", null),
            billingCycle = obj.optString("billingCycle", null),
            lastPaymentAmount = obj.optString("lastPaymentAmount", null),
            lastActiveDate = obj.optString("lastActiveDate", null),
            totalSessions = obj.optString("totalSessions", null),
            custom = parseCustomMap(obj),
        )
    }

    private fun parseCustomMap(obj: org.json.JSONObject): Map<String, String> {
        val customObj = obj.optJSONObject("custom") ?: return emptyMap()
        val result = mutableMapOf<String, String>()
        val keys = customObj.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = customObj.optString(key, null)
            if (value != null) {
                result[key] = value
            }
        }
        return result
    }
}
